go-controller/pkg/ovn/controller/egressfirewall/egressfirewall.go

```diff
@@ -374,44 +397,73 @@ func (oc *EFController) sync(key string) (updateErr error) {
 	if err != nil {
 		return fmt.Errorf("invalid resource key for egress firewall: %s", key)
 	}
-	shouldRemove := false
-	shouldAddUpdate := false
+	oc.cache.LockKey(namespace)
+	defer oc.cache.UnlockKey(namespace)
+
+	existingEntry, _ := oc.cache.Load(namespace)
+	var newEntry *cacheEntry
 
+	efExists := false
 	ef, err := oc.efLister.EgressFirewalls(namespace).Get(efName)
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return err
 		}
-		shouldRemove = true
 	} else {
-		shouldAddUpdate = true
+		efExists = true
+		activeNetwork, netErr := oc.networkManager.GetActiveNetworkForNamespace(namespace)
+		if netErr != nil {
+			if util.IsInvalidPrimaryNetworkError(netErr) {
+				// Namespace requires P-UDN, but it does not exist. Remove EF config and surface error in status.
+				updateErr = netErr
+			} else {
+				return fmt.Errorf("failed to get active network for egress firewall %s/%s namespace: %w",
+					namespace, efName, netErr)
+			}
+		} else {
+			ownerController := activeNetwork.GetNetworkName() + "-network-controller"
+			newEntry = &cacheEntry{
+				pgName:            libovsdbutil.GetPortGroupName(getNamespacePortGroupDbIDs(namespace, ownerController)),
+				subnetsKey:        subnetsKeyForNetInfo(activeNetwork),
+				efResourceVersion: ef.ResourceVersion,
+			}
+		}
 	}
 
-	oc.cache.LockKey(namespace)
-	defer oc.cache.UnlockKey(namespace)
+	if efExists {
+		defer func() {
+			if statusErr := oc.setEgressFirewallStatus(ef, updateErr); statusErr != nil {
+				updateErr = utilerrors.Join(updateErr, fmt.Errorf("failed to update egress firewall status %s, error: %w",
+					GetEgressFirewallNamespacedName(ef), statusErr))
+			}
+		}()
+	}
 
-	if entry, ok := oc.cache.Load(namespace); ok {
-		if entry.stale {
-			shouldRemove = true
-		}
+	// If nothing relevant changed since last apply, skip update.
+	if entriesEqual(existingEntry, newEntry) {
+		return updateErr
 	}
 
-	if shouldRemove {
-		// delete case
-		entry, ok := oc.cache.Load(namespace)
-		if !ok {
-			return nil
+	// Delete existing state if there is no desired state.
+	if newEntry == nil {
```

Comment:
should we also do this if port gorup name has changed? then you could remove the extra cleanup in the end of this function

Reply:
maybe skip dnsNameresolved cleanup if newEntry != nil?